### ISSUE TITLE
softplus overflows to inf for large inputs, producing NaN gradients

### ISSUE THREAD

**Issue by Mikyx-1** [tracel-ai/burn#5373]
**Describe the bug**

`activation::softplus` computes `exp` before `log`:

https://github.com/tracel-ai/burn/blob/main/crates/burn-tensor/src/tensor/activation/base.rs#L229-L232

```rust
let tensor = (tensor.mul_scalar(beta).exp() + 1).log();
tensor.div_scalar(beta)
```

`exp` overflows above ~88.7 in `f32`, so `log(inf) = inf`. Because `beta` scales the input, a `beta` of 5 overflows from `x ≈ 18` upwards — well inside a normal activation range.

The forward `inf` is the visible symptom, but the backward pass is the more serious one: the saturated element yields a **`NaN` gradient**, which then propagates into the shared parameter gradients and corrupts the whole update step, not just that element.

`SoftMarginLoss` inherits this. It is defined as `softplus(-target * logit)`, and a confidently *wrong* prediction is exactly the case that saturates — so the loss returns `inf` and the backward pass returns `NaN` precisely when the model is most wrong.

(`mish` also calls `softplus`, but happens to survive: `tanh(inf) = 1`, so `mish(x) -> x`, which is the correct limit.)

**To Reproduce**

Measured on the `ndarray` backend, `f32`:

```rust
// forward
softplus([20, 50, 100], beta = 1.0)   ->  [20.0, 50.0, inf]      // expected [20.0, 50.0, 100.0]
softplus([20, 50, 100], beta = 5.0)   ->  [inf,  inf,  inf]      // expected [20.0, 50.0, 100.0]

// SoftMarginLoss, logits [-100, -20, -5] with targets [1, 1, 1]
->  [inf, 20.0, 5.0067153]                                        // expected [100.0, 20.0, 5.0067153]

// backward
softplus([-100, 0, 100], beta = 1.0).sum().backward()
->  grad [3.8e-44, 2.06e-9, 0.5, 1.0, 1.0, NaN]                   // expected sigmoid(x), no NaN
```

There is a second, quieter issue in the same expression. The `1 + exp(beta * x)` intermediate rounds to exactly `1.0` in `f32` once `beta * x < ~-16`, so the negative tail underflows to zero:

```rust
softplus([-20, -50], beta = 1.0)  ->  [0.0, 0.0]   // expected [2.06e-9, 1.93e-22]
```

**Expected behavior**

Softplus should saturate to the identity for large `beta * x` and to `exp(beta * x) / beta` for very negative `beta * x`, with a finite `sigmoid(beta * x)` gradient throughout. PyTorch guards this with a `threshold` parameter (default `20`), above which it returns the input linearly; `burn`'s `softplus` has no such parameter.

Existing coverage in `crates/burn-backend-tests/tests/tensor/float/activation/softplus.rs` only tests values in `[-1, 1]`, and there is no autodiff test for `softplus` at all, which is why neither the `inf` nor the `NaN` was caught.

Notably, `burn` already handles this exact class of problem for `log_sigmoid`, using the log-sum-exp trick with an explanatory comment:

https://github.com/tracel-ai/burn/blob/main/crates/burn-backend/src/backend/ops/activation.rs#L201-L217

So this looks like an oversight in `softplus` rather than a deliberate tradeoff.

**Additional context**

- Backend: `ndarray` (the bug is in the backend-agnostic `burn-tensor` composition, so it affects all backends)
- Dtype: `f32`
- OS: macOS
